### DISCUSSION TITLE
[BUG] shebang (#!) maximum line length limit 128 in ubuntu-latest workers instead of 256

### DISCUSSION THREAD

**Discussion by jens-maus** [community/community#26222] [Actions]
<p>Hello,</p>
<p>I am sucessfully using GitHub Actions in one of my open source projects, but I think I came across a small, but important issue in the “ubuntu-latest” worker environment which prevent one of my build jobs to sucessfully build. In this project I run long build jobs and build a complete OS image based on the <a href="http://buildroot.org" rel="nofollow noopener">buildroot.org</a> linux distribution.</p>
<p>As said, for one of my matrix jobs the build fails and after some investigation I found the root cause of the issue being that within the “ubuntu-latest” runner/worker environment the shebang maximum line length limit (BINPRM_BUF_SIZE in /usr/include/linux/binfmts.h) seem to be 128 in the ubuntu-latest 18.04 workers, while in a standard ubuntu 19.10 environment the shebang limit is set to 256. See here:</p>
<p>Standard Ubuntu 19.10:</p>
<pre><code>$ grep BINPRM_BUF_SIZE /usr/include/linux/binfmts.h
#define BINPRM_BUF_SIZE 256
</code></pre>
<p>“ubuntu-latest” worker in GitHub Actions:</p>
<pre><code>runner@fv-az32:$grep BINPRM\_BUF\_SIZE /usr/include/linux/binfmts.h#define BINPRM_BUF_SIZE 128
</code></pre>
<p>As a result, shell scripts using the shebang qualifier (#! in the top-line of the script) are limited to path length of 128 characters rather than 256 which seem to be more sufficient these days, especially since the checked out directory on a worker is located in /home/runner/work/&lt;PROJECTNAME&gt;/&lt;PROJECTNAME&gt;, which can easily make up already quite a portion of the maximum allowed path/shebang line length limit. In my case this standard path already sums up to 48 out of 128 max characters (/home/runner/work/RaspberryMatic/RaspberryMatic).</p>
<p>So here, in one of the workers the following shell script cannot be executed successfully on the GitHub Actions “ubuntu-latest” workers:</p>
<pre><code>#!/home/runner/work/RaspberryMatic/RaspberryMatic/build-raspmatic_ova/build/recovery-system-1.7.1-2019.11.1/output/host/bin/python3
# EASY-INSTALL-ENTRY-SCRIPT: 'meson==0.52.0','console_scripts','meson'
__requires__ = 'meson==0.52.0'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == ' __main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('meson==0.52.0', 'console_scripts', 'meson')()
    )
</code></pre>
<p>And that’s because the shebang at the top-line sums up to 132 characters. </p>
<p>After reducing the directory path length by renaming the last “RaspberryMatic” directory to a shortcut name the script can be successfully executed and the build runs correctly.</p>
<p>However, I still feel that the shebang limit should be increased to 256 characters like this already seem to be the case for newer ubuntu versions as otherwise build jobs like mine (which use a third party build environment with automatically generated pathes) may run into the same issue.</p>

**Comment by BrightRan** [comment]
<p>Looks like this is an OS limitation, not a virtual environment problem.</p>
<p>Currently the latest Linux virtual environment for GitHub-hosted runner is Ubuntu 18.04, I recommend you can try using a <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/hosting-your-own-runners" rel="nofollow noopener"><strong>self-hosted runner</strong></a> which is installed on your local machine (or VM) with OS is Ubuntu 19.10 or later.</p>
<p>There is a similar case as reference: <a href="https://github.com/pypa/virtualenv/issues/596" rel="nofollow noopener">https://github.com/pypa/virtualenv/issues/596</a></p>

**Comment by jens-maus** [comment, marked as answer]
<p>Please note that I worked around this OS limitation issue within GitHub Actions by having added the following step snippet right after the checkout:</p>
<pre><code>steps:
    - uses: actions/checkout@v2

    - name: rename worker directory
      working-directory: ../
      run: |
        mv RaspberryMatic b
        ln -s b RaspberryMatic
</code></pre>
<p>Thus, I am simply renaming my working directory from “RaspberryMatic” to “b” and added a symbolic link. This has the effect that during the buildroot build this drops the shebang length limit below 128 and thus the build finishes correctly.</p>
